### PR TITLE
fix(ui): roll back an optimistic reaction edit when the backend rejects it (#615)

### LINKED ISSUE #615: Reaction appears updated on UI without proper permissions after role is changed to viewer
## Preconditions
1. Group is created
2. The second admin is added to Group
3. Upload dataset with reaction to this Group 
Image
4. Sign in as admin 1 in 1 browser
5. Sign in as admin 2 in 2 browser
6. As admin 2 open the reaction in dataset
7. As admin 1, change the role of admin 2 to Viewer
8. As admin 2, not refresh the page, try to update the reaction (e.g.: delete input and output)

## Actual Result 
- BE issues are in console and error message is on UI, but Reaction is updated on UI:
Image
- If user refreshes the page, it's not updated:
Image

## Expected Result 
- No updates on UI, as user doesn't have access to update reaction

## Environment 
- Reproduced in the local environment with main branch, commit ed13c19c1c099e4a32aa33ccee958c588a976403;
- dev-z4acb31kcl4prqtw.us.auth0.com auth tenant;
- Desktop, Windows 11 OS, Chrome Browser.

### PR DESCRIPTION
**Pull request by skearnes**
## Summary

Cluster D (permissions/role-change reactivity) — the optimistic-update half. Companion to #770 (re-gate + notifications); this one stands alone.

Reaction field edits are **optimistic**: the `.request` reducer mutates the reaction's local `data` immediately, which is load-bearing — the thunk then reads that mutated state to encode the protobuf for the PATCH. But there was **no `.failure` handler**, so when the backend rejected the edit, the optimistic change stuck:

- **Case 1** (#615): role changed to viewer → PATCH 403s → an error toast appears, but the UI still shows the edit (e.g. a deleted input stays gone) until a manual reload.
- **Case 2** (#615): backend stopped → request fails → same thing, the user sees changes that were never saved.

### Fix
Snapshot the pre-edit `data` into a transient `dataBeforeEdit` on the add/update and delete field `.request`, and **restore it on `.failure`**:
- Keeps the **earliest** baseline if several edits queue up, so a rollback reverts all the way.
- Implicitly cleared on `.success` — the entry is rebuilt from the server payload, which carries no snapshot.
- Edits are issued one at a time (each dispatch is awaited), so at most one reaction is ever pending; the failure handler restores any snapshotted entry.

This is purely a reducer/type change — the thunks and action payloads are untouched. It works for **both** rejection modes (403 and a no-response network failure), since both dispatch `.failure`.

## Test plan
- [x] `tsc -b` clean; `prettier` / `eslint` clean
- [x] **689 unit tests pass**, including 6 new reducer cases in `reactions.reducer.test.ts`:
  - snapshots the pre-edit data on a field edit
  - restores the snapshot and clears it on `.failure` (the bug)
  - clears the snapshot on `.success` (committed edit isn't reverted)
  - keeps the earliest baseline across successive edits
  - `.failure` is a no-op when nothing is pending
- [x] **Runtime-verified on the live no-auth stack** (provisioned a 2nd user + toggled the logged-in user's group role; see verification comment): on a rejected edit (`403`) the optimistic change reverts (entity reappears) — vs `main`, where it persists. The reducer logic is also fully unit-tested, and the "backend down" case (Case 2) by the failure-path tests.

Closes #615.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR adds a rollback mechanism for optimistic reaction field edits: when the backend rejects a PATCH (403 on role change, or network failure), the UI now restores the pre-edit state instead of leaving the user with changes that were never persisted.

- Snapshots `data` into `dataBeforeEdit` on each `.request`, preserving the earliest baseline across queued edits so a rollback always reverts to the true pre-flight state.
- Restores and clears the snapshot on `.failure`; explicitly clears it on `.success`; `UpdateReactionSuccessPayload` now explicitly omits `dataBeforeEdit` so a future payload carrying the field cannot accidentally keep the snapshot alive after a committed edit.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is confined to the reducer and its types, the happy path is unchanged, and all six new test cases pass.

The fix is purely additive state logic: new fields on an existing interface, two new lines in the `.request` handlers, an explicit clear on `.success`, and a new failure matcher. Both previously flagged gaps (implicit snapshot clearing via server payload and the `UpdateReactionSuccessPayload` type hole) are addressed directly in this PR. The failure handler correctly skips entries with no snapshot and returns the original reference when nothing was rolled back, so there is no unnecessary re-render. The test suite covers all meaningful states.

No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| ui/src/store/entities/reactions/reactions.types.ts | Adds `dataBeforeEdit?: AppReaction` to `BaseReaction`; explicitly omits it from `UpdateReactionSuccessPayload` to prevent the server response from accidentally carrying the client-only snapshot. |
| ui/src/store/entities/reactions/reactions.reducer.ts | Snapshot logic added to both `.request` handlers; failure matcher restores snapshot and clears it; success handler explicitly clears `dataBeforeEdit`. Logic is sound and handles edge cases (earliest-baseline preservation, no-op when nothing is pending). |
| ui/src/store/entities/reactions/reactions.reducer.test.ts | Six new reducer test cases covering snapshot creation, rollback on failure, clear on success, earliest-baseline preservation, and no-op failure; coverage is comprehensive for the new logic. |

</details>

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `ui/src/store/entities/reactions/reactions.types.ts`, line 168 ([link](https://github.com/open-reaction-database/ord-app/blob/bf6cc9c37a18202ded7de7c97b6790061b3f9eb7/ui/src/store/entities/reactions/reactions.types.ts#L168)) 

   <a href="#">P2</a> `UpdateReactionSuccessPayload` is `Omit<DatasetReaction, 'data'>`, and `DatasetReaction` now inherits `dataBeforeEdit?` from `BaseReaction`. So the success payload type quietly includes `dataBeforeEdit?: AppReaction`, which the server never actually returns. TypeScript will happily accept a success payload that carries a snapshot, and the success handler would spread it into the new entry — keeping the snapshot alive after the edit was committed. Omitting `dataBeforeEdit` from the payload type closes that gap.
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(ui): clear edit snapshot explicitly ..."](https://github.com/open-reaction-database/ord-app/commit/2f0c52b5da34e77680dde3d5bbca9c9966a19625) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38751516)</sub>

### CODE REVIEW

File: ui/src/store/entities/reactions/reactions.reducer.ts [hunk 1]
```diff
@@ -173,6 +181,21 @@ const reactionsById = createReducer<ItemsById<ReactionOrTemplate>>({}, builder =
       };
```
[the remark this answers is not in the document]
Comment (skearnes) [reply ui/src/store/entities/reactions/reactions.reducer.ts:182]: Good call on both — addressed in the latest commit. The success handler now sets `dataBeforeEdit: undefined` explicitly, and `UpdateReactionSuccessPayload` is `Omit<DatasetReaction, 'data' | 'dataBeforeEdit'>` so the type reflects that the server never returns the client-only snapshot.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by skearnes** [comment]
✅ **Runtime-verified on the live no-auth stack** with a before/after contrast (provisioned a 2nd user + toggled the logged-in user's group role to viewer mid-session; confirmed a downgraded viewer gets `403` on writes even on their own dataset).

**Scenario** (Playwright): open a reaction as admin, downgrade to viewer, delete an entity → the optimistic removal applies, then the `PATCH` is rejected.

| | entities before | after optimistic delete | after `403 PATCH` | rolled back? |
|---|---|---|---|---|
| **this branch (#771)** | 9 | 8 | **9** | ✅ yes |
| **`main` (no fix)** | 9 | 8 | **8** | ❌ no — the unsaved delete persists (the bug) |

So with the fix the rejected edit is reverted (entity reappears); on `main` it sticks until a manual refresh — exactly #615. An "Access forbidden" toast is shown in both cases. (Also exercises the same 403 path that #770 re-gates on.)

_Analysis performed with [Claude Code](https://claude.com/claude-code) (Claude Opus 4.8) — verified against a live local stack._
